design/client-policies.md

```diff
@@ -0,0 +1,191 @@
+
+However, it can only cover OIDC Dynamic Client Registration.
+
+Client Policies cover not only what Client Registration Policies can do but other client registration and configuration ways.
+
+In the future, Client Registration will be replaced by Client Policies.
+
+### 2. Validation of client configurations
+
+The current keycloak has already supported validation on whether the client follows settings like Proof Key for Code Exchange, Request Object Signing Algorithm, Holder-of-Key Token, etc on some endpoints like Authorization Endpoint, Token Endpoint, etc.
+
+These can be specified by each setting item (on Admin Console, switch, pulldown menu, etc). To make the client application secure, user need to set a lot of such settings in appropriate way, which makes it difficult for the user to secure their client application.
+
+Client Policies can do these validation of client configurations mentioned just above. In the futre, these each setting items will be replaced by Client Policies. 
+
+The pre-set client policies for following some security profiles (FAPI, SPA, Native App, etc) are provided, which makes it easy for the user to secure their client application.
+
+### 3. Conformance to a profile such as FAPI
+
+It is treated in [Client Conformance Profiles](https://issues.redhat.com/browse/KEYCLOAK-11612) based on this Client Policies. Which kind of security profiles are supported and how to realize them are discussed on 
+[Client Conformance Profiles](https://issues.redhat.com/browse/KEYCLOAK-11612).
+
+## Implementation details
+
+Client Policies consists of the three building blocks, Policy, Condition, and Executor.
+
+### Condition : which client
+
+A condition determins to which client a policy is adopted. The condition checks whether one specified criteria is satisfied. Such the task is that checking whether the [access type](https://www.keycloak.org/docs/latest/server_admin/index.html#oidc-clients) of the client is confidential.
+
+#### How to implement
+
+A condition can be implemented as a provider (`ClientPolicyConditionSpi`, `ClientPolicyConditionFactory`, `ClientPolicyCondition`). Developers can add their own condition by implementing this provider.
+
+#### How to configure
+
+A condition can be configuable the same as other configurable provider. What can be configured depends on each condition's nature.
+
+#### Which kind of conditions are provided
+
+The following conditions are provided:
+
+* The way of creating/updating a client (OIDC Dynamic Client Registration - Authenticated, Anonymous, Admin REST API - Admin Console, etc.)
+
+E.g. When creating a client, a policy is adopted when this client is created by OIDC Dynamic Client Registration with Initial Access Token.
+
+* Client Access Type (confidential, public, bearer-only)
+
+E.g. When a client sends an authorization request, a policy is adopted if this client is confidential.
+
+* Client ID
```

Comment:
Not sure I see how "Client ID" is that useful as a condition. It's to specific and would make it very hard to manage policies. Client ID is also not something that is controllable - to prevent a policy being triggered a developer could just use a different client-id.

Reply:
I see, I'll remove this "Client ID" condition.